src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,692 @@
+      for (size_t k = 0; k < pcol; ++k) Assert(reslhs[j][k] == 0);
+#endif
+      /* normalize element in pivot column to modulo prime */
+      reslhs[j][pcol] = reslhs[j][pcol].euclidianDivideRemainder(prime);
+      /* exchange rows if pivot elem is 0 */
+      if (j == prow)
+      {
+        while (reslhs[j][pcol] == 0)
+        {
+          for (size_t k = prow + 1; k < nrows; ++k)
+          {
+            reslhs[k][pcol] = reslhs[k][pcol].euclidianDivideRemainder(prime);
+            if (reslhs[k][pcol] != 0)
+            {
+              std::swap(resrhs[j], resrhs[k]);
+              std::swap(reslhs[j], reslhs[k]);
+              break;
+            }
+          }
+          if (pcol >= ncols - 1) break;
+          if (reslhs[j][pcol] == 0)
+          {
+            pcol += 1;
+            if (reslhs[j][pcol] != 0)
+              reslhs[j][pcol] = reslhs[j][pcol].euclidianDivideRemainder(prime);
+          }
+        }
+      }
+      /* (1) */
+      if (reslhs[j][pcol] != 0)
+      {
+        if (reslhs[j][pcol] != 1)
+        {
+          Integer inv = reslhs[j][pcol].modInverse(prime);
+          if (inv == -1)
+          {
+            return BVGaussElim::Result::INVALID; /* not coprime */
+          }
+          for (size_t k = pcol; k < ncols; ++k)
+          {
+            reslhs[j][k] = reslhs[j][k].modMultiply(inv, prime);
+            if (j <= prow) continue; /** pivot */
+            reslhs[j][k] = reslhs[j][k].modAdd(-reslhs[prow][k], prime);
+          }
+          resrhs[j] = resrhs[j].modMultiply(inv, prime);
+          if (j > prow) resrhs[j] = resrhs[j].modAdd(-resrhs[prow], prime);
+        }
+        /* (2) */
+        else if (j != prow)
+        {
+          for (size_t k = pcol; k < ncols; ++k)
+            reslhs[j][k] = reslhs[j][k].modAdd(-reslhs[prow][k], prime);
+          resrhs[j] = resrhs[j].modAdd(-resrhs[prow], prime);
+        }
+      }
+    }
+    /* (3) */
+    for (size_t j = 0; j < prow; ++j)
+    {
+      Integer mul = reslhs[j][pcol];
+      if (mul != 0)
+      {
+        for (size_t k = pcol; k < ncols; ++k)
+          reslhs[j][k] = reslhs[j][k].modAdd(-reslhs[prow][k] * mul, prime);
+        resrhs[j] = resrhs[j].modAdd(-resrhs[prow] * mul, prime);
+      }
+    }
+  }
+
+  bool ispart = false;
+  for (size_t i = 0; i < nrows; ++i)
+  {
+    size_t pcol = i;
+    while (pcol < ncols && reslhs[i][pcol] == 0) ++pcol;
+    if (pcol >= ncols)
+    {
+      resrhs[i] = resrhs[i].euclidianDivideRemainder(prime);
+      if (resrhs[i] != 0) return BVGaussElim::Result::NONE; /* no solution */
+      continue;
+    }
+    for (size_t j = i; j < ncols; ++j)
+    {
+      if (reslhs[i][j] >= prime || reslhs[i][j] <= -prime)
+        reslhs[i][j] = reslhs[i][j].euclidianDivideRemainder(prime);
+      if (j > pcol && reslhs[i][j] != 0) ispart = true;
+    }
+  }
+  if (ispart) return BVGaussElim::Result::PARTIAL;
+
+  return BVGaussElim::Result::UNIQUE;
+}
+
+static bool is_bv_const(Node n)
```

Comment:
`isBVGroundTerm` can not be applied here. However, I refactored my approach / algorithm to locally apply rewriting for more robustness. Previously, `x*y` and `y*x` where interpreted as two different unknowns, resulting in two columns in the matrix. In my upcoming changes (not pushed yet) this will be recognized as being equivalent. Also, arbitrary constant expressions will be simplified, e.g., `2 + 3` will be recognized as `5`, which was not the case previously.